File: sentry/src/test/java/io/sentry/servlet/SentryServletRequestListenerTest.java
package io.sentry.servlet;

import io.sentry.BaseTest;
import org.junit.Before;
import org.junit.Test;

import javax.servlet.ServletRequest;
import javax.servlet.ServletRequestEvent;
import javax.servlet.http.HttpServletRequest;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.nullValue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class SentryServletRequestListenerTest extends BaseTest {
    private SentryServletRequestListener sentryServletRequestListener = null;
    private ServletRequestEvent mockServletRequestEvent = null;

    @Before
    public void setup() {
        mockServletRequestEvent = mock(ServletRequestEvent.class);
        sentryServletRequestListener = new SentryServletRequestListener();
        SentryServletRequestListener.reset();
    }

    @Test
    public void requestListenerEmptyByDefault() throws Exception {
        assertThat(SentryServletRequestListener.getServletRequest(), is(nullValue()));
    }

    @Test
    public void requestListenerContainsTheCurrentRequest() throws Exception {
        HttpServletRequest mockHttpServletRequest = mock(HttpServletRequest.class);
        when(mockServletRequestEvent.getServletRequest()).thenReturn(mockHttpServletRequest);

        sentryServletRequestListener.requestInitialized(mockServletRequestEvent);

        assertThat(SentryServletRequestListener.getServletRequest(), is(mockHttpServletRequest));
    }

    @Test
    public void requestListenerDoesntWorkWithNonHttpRequests() throws Exception {
        ServletRequest mockServletRequest = mock(ServletRequest.class);
        when(mockServletRequestEvent.getServletRequest()).thenReturn(mockServletRequest);

        sentryServletRequestListener.requestInitialized(mockServletRequestEvent);

        assertThat(SentryServletRequestListener.getServletRequest(), is(nullValue()));
    }

    @Test
    public void requestListenerDestroyRemovesTheCurrentRequest() throws Exception {
        HttpServletRequest mockHttpServletRequest = mock(HttpServletRequest.class);
        when(mockServletRequestEvent.getServletRequest()).thenReturn(mockHttpServletRequest);

        sentryServletRequestListener.requestInitialized(mockServletRequestEvent);

        sentryServletRequestListener.requestDestroyed(mockServletRequestEvent);

        assertThat(SentryServletRequestListener.getServletRequest(), is(nullValue()));
    }

    @Test
    public void requestListenerSpecificToLocalThread() throws Exception {
        HttpServletRequest mockHttpServletRequest = mock(HttpServletRequest.class);
        when(mockServletRequestEvent.getServletRequest()).thenReturn(mockHttpServletRequest);

        new Thread() {
            @Override
            public void run() {
                sentryServletRequestListener.requestInitialized(mockServletRequestEvent);
            }
        }.start();

        assertThat(SentryServletRequestListener.getServletRequest(), is(nullValue()));
    }
}

File: sentry-servlet/src/test/kotlin/io/sentry/servlet/SentryServletRequestListenerTest.kt
package io.sentry.servlet

import io.sentry.Breadcrumb
import io.sentry.IScopes
import io.sentry.ISentryLifecycleToken
import org.assertj.core.api.Assertions.assertThat
import org.mockito.kotlin.any
import org.mockito.kotlin.anyOrNull
import org.mockito.kotlin.check
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import org.springframework.mock.web.MockHttpServletRequest
import javax.servlet.ServletRequestEvent
import kotlin.test.Test
import kotlin.test.assertSame

class SentryServletRequestListenerTest {
    private class Fixture {
        val scopes = mock<IScopes>()
        val lifecycleToken = mock<ISentryLifecycleToken>()
        val listener = SentryServletRequestListener(scopes)
        val request = MockHttpServletRequest()
        val event = mock<ServletRequestEvent>()

        init {
            request.requestURI = "http://localhost:8080/some-uri"
            request.method = "post"
            whenever(event.servletRequest).thenReturn(request)
            whenever(scopes.forkedScopes(any())).thenReturn(scopes)
            whenever(scopes.makeCurrent()).thenReturn(lifecycleToken)
        }
    }

    private val fixture = Fixture()

    @Test
    fun `pushes scope when request gets initialized`() {
        fixture.listener.requestInitialized(fixture.event)

        verify(fixture.scopes).forkedScopes(any())
        verify(fixture.scopes).makeCurrent()
    }

    @Test
    fun `adds breadcrumb when request gets initialized`() {
        fixture.listener.requestInitialized(fixture.event)

        verify(fixture.scopes).addBreadcrumb(
            check { it: Breadcrumb ->
                assertThat(it.getData("url")).isEqualTo("http://localhost:8080/some-uri")
                assertThat(it.getData("method")).isEqualTo("POST")
                assertThat(it.type).isEqualTo("http")
            },
            anyOrNull(),
        )
        assertSame(fixture.lifecycleToken, fixture.request.getAttribute("sentry-scope-lifecycle"))
    }

    @Test
    fun `pops scope when request gets destroyed`() {
        fixture.request.setAttribute("sentry-scope-lifecycle", fixture.lifecycleToken)

        fixture.listener.requestDestroyed(fixture.event)
        verify(fixture.lifecycleToken).close()
    }
}
